mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java

```diff
@@ -10,293 +10,329 @@
-            Log.e("OpenHABWidget", "Color was " + color);
-            Log.e("OpenHABWidget", e.getMessage());
-            this.iconcolor = null;
+    private static String determineOH2IconPath(OpenHABItem item, Type type, String icon,
+            String iconFormat, boolean hasMappings) {
+        String itemState = item != null ? item.state() : null;
+        if (itemState != null) {
+            if (item.isOfTypeOrGroupType(OpenHABItem.Type.Color)) {
+                // For items that control a color item fetch the correct icon
+                if (type == Type.Slider || (type == Type.Switch && !hasMappings)) {
+                    try {
+                        itemState = String.valueOf(item.stateAsBrightness());
+                        if (type == Type.Switch) {
+                            if (itemState.equals("0")) {
+                                itemState = "OFF";
+                            } else {
+                                itemState = "ON";
+                            }
+                        }
+                    } catch (Exception e) {
+                        itemState = "OFF";
+                    }
+                }
+            } else if (type == Type.Switch && !hasMappings
+                    && !item.isOfTypeOrGroupType(OpenHABItem.Type.Rollershutter)) {
+                // For switch items without mappings (just ON and OFF) that control a dimmer item
+                // set the state to "OFF" instead of 0 or to "ON" to fetch the correct icon
+                try {
+                    int itemStateNumber = Integer.valueOf(itemState);
+                    if (itemStateNumber == 0) {
+                        itemState = "OFF";
+                    } else {
+                        itemState = "ON";
+                    }
+                } catch (java.lang.NumberFormatException e) {
+                    // Item state is not a number, not sure if that can happen, but good to catch
+                }
+            }
         }
-    }
 
-    private String fixColorName(String colorName) {
-        if (colorName.equals("orange"))
-            return "#FFA500";
-        return colorName;
+        return String.format("icon/%s?state=%s&format=%s", icon, itemState, iconFormat);
     }
 
-    public String getEncoding() {
-        return encoding;
+    private static Type parseType(String type) {
+        if (type != null) {
+            try {
+                return Type.valueOf(type);
+            } catch (IllegalArgumentException e) {
+                // fall through
+            }
+        }
+        return Type.Unknown;
     }
 
-    public void setEncoding(String encoding) {
-        this.encoding = encoding;
+    private static Integer parseColor(String color) {
+        if (color != null && !color.isEmpty()) {
+            if (color.equals("orange")) {
+                color = "#FFA500";
+            }
+            try {
+                return new Integer(Color.parseColor(color));
```

Comment:
Android Studio suggests to change that to `Integer.valueOf(Color.parseColor(color))`.

Reply:
Done.